hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/StoreSnapshotTasklet.java

```diff
@@ -79,64 +83,75 @@ private void stateMachineStep() {
+                        if (!ssWriter.offer((Entry<Data, Data>) o)) {
+                            pendingEntry = (Entry<Data, Data>) o;
+                            return false;
+                        }
                     }
+                    return true;
                 });
                 if (result.isDone()) {
+                    state = REACHED_BARRIER;
+                    assert ssWriter.isEmpty() : "input is done, but we had some entries and not the barrier";
                     inputIsDone = true;
                 }
-                if (result.isMadeProgress()) {
-                    progTracker.madeProgress();
-                    state = FLUSH;
+                progTracker.madeProgress(result.isMadeProgress());
+                if (hasReachedBarrier) {
+                    state = FLUSH_REMAINING;
                     stateMachineStep();
                 }
                 return;
 
-            case FLUSH:
+            case FLUSH_REMAINING:
                 progTracker.notDone();
-                CompletableFuture<Void> future = new CompletableFuture<>();
-                future.whenComplete(withTryCatch(logger, (r, t) -> {
-                    // this callback may be called from a non-tasklet thread
-                    if (t != null) {
-                        logger.severe("Error writing to snapshot map '" + currMapName() + "'", t);
-                        snapshotContext.reportError(t);
-                    }
-                    // numActiveFlushes must be decremented last otherwise we may miss the error
-                    numActiveFlushes.decrementAndGet();
-                }));
-                if (mapWriter.tryFlushAsync(future)) {
+                if (ssWriter.flushRemaining()) {
                     progTracker.madeProgress();
-                    numActiveFlushes.incrementAndGet();
-                    state = inputIsDone ? DONE : hasReachedBarrier ? REACHED_BARRIER : DRAIN;
+                    state = REACHED_BARRIER;
                 }
                 return;
 
             case REACHED_BARRIER:
-                progTracker.notDone();
-                if (numActiveFlushes.get() == 0) {
+                boolean success = false;
+                try {
+                    success = ssWriter.hasPendingFlushes();
+                } catch (Throwable e) {
+                    logger.severe("Error writing to snapshot map '" + currMapName() + "'", e);
+                    snapshotContext.reportError(e);
+                }
+                if (success) {
+                    progTracker.notDone();
+                } else {
+                    progTracker.madeProgress();
                     snapshotContext.snapshotDoneForTasklet();
                     pendingSnapshotId++;
-                    mapWriter.setMapName(currMapName());
-                    state = inputIsDone ? DONE : DRAIN;
+                    resetCurrentMap();
                     hasReachedBarrier = false;
+                    if (inputIsDone) {
```

Comment:
I think it was better with two explicit states. The code is hard to follow.

Reply:
fixed